src/socket/raw.cr

```diff
@@ -0,0 +1,693 @@
+require "c/sys/socket"
+require "./addrinfo"
+
+# This class represents a raw network socket.
+#
+# It is an object oriented wrapper for BSD-style socket API provided by POSIX operating
+# systems and Windows.
+#
+# This class is not intended to be used for typical network applications. There
+# are more specific implementations `TCPSocket`, `UDPSocket`, `UNIXSocket`, `TCPServer`, and `UNIXServer`.
+# It allows finer-grained control over socket parameters than the protocol-specific classes
+# and only needs to be employed for less common tasks that need low-level access to the OS sockets.
+class Socket::Raw < IO
+  include IO::Buffered
+  include IO::Syscall
+
+  # The raw file-descriptor. It is defined to be an `Int32`, but its actual size is
+  # platform-specific.
+  getter fd : Int32
+
+  @read_event : Crystal::Event?
+  @write_event : Crystal::Event?
+
+  @closed : Bool
+
+  getter family : Family
+  getter type : Type
+  getter protocol : Protocol
+
+  # Creates a new raw socket for TCP protocol.
+  #
+  # Consider using `TCPSocket` or `TCPServer` instead.
+  def self.tcp(family : Family, *,
+               blocking : Bool = false)
+    new(family, Type::STREAM, Protocol::TCP, blocking: blocking)
+  end
+
+  # Creates a new raw socket for UDP protocol.
+  #
+  # Consider using `UDPSocket` instead.
+  def self.udp(family : Family, *,
+               blocking : Bool = false)
+    new(family, Type::DGRAM, Protocol::UDP, blocking: blocking)
+  end
+
+  # Creates a new raw socket for UNIX sockets.
+  #
+  # Consider using `UNIXSocket` or `UNIXServer` instead.
+  def self.unix(type : Type = Type::STREAM, *,
+                blocking : Bool = false)
+    new(Family::UNIX, type, blocking: blocking)
+  end
+
+  # Creates a new raw socket.
+  def initialize(@family : Family, @type : Type, @protocol : Protocol = Protocol::IP, *,
+                 blocking : Bool = false)
+    @closed = false
+    fd = LibC.socket(family, type, protocol)
+    raise Errno.new("failed to create socket:") if fd == -1
+    init_close_on_exec(fd)
+    @fd = fd
+
+    self.sync = true
+    unless blocking
+      self.blocking = false
+    end
```

Comment:
I would rather use `def self.new`, remove the last 5 duplicated lines along with `@closed = true`, and at the end call the protected initializer below `new(fd, family, type, protocol, blocking)`